pmd-core/src/main/java/net/sourceforge/pmd/renderers/XMLRenderer.java

```diff
@@ -179,6 +182,41 @@ private void createTimestampAttr(StringBuilder buffer) {
                 .append('"');
     }
 
+    @Override
+    public void setReportFile(String reportFilename) {
+        String encoding = getProperty(ENCODING);
+
+        try {
+            Charset charset = Charset.forName(encoding);
+            this.writer = StringUtils.isBlank(reportFilename) ? new OutputStreamWriter(System.out, charset)
+                    : Files.newBufferedWriter(new File(reportFilename).toPath(), charset);
+        } catch (IOException | UnsupportedCharsetException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Escape unicode characters for non UTF-8 encodings.
+     */
+    private String escape(String text) {
+        String result = StringEscapeUtils.escapeXml10(text);
+        String encoding = getProperty(ENCODING);
+        if (!"UTF-8".equalsIgnoreCase(encoding)) {
+            StringBuilder sb = new StringBuilder(result);
+            for (int i = 0; i < sb.length(); i++) {
+                char c = sb.charAt(i);
+                // surrogate characters are not allowed in XML
```

Comment:
this and a non unicode based encoding like ISO-8859-1 or cp1252 cannot encode codepoints > 255.
Hm... so, we should probably escape chars additionally as soon as `c > 0xff`, because then you need unicode...